x/ccv/consumer/keeper/genesis.go

```diff
@@ -37,39 +42,51 @@ func (k Keeper) InitGenesis(ctx sdk.Context, state *consumertypes.GenesisState)
 	}
 
 	// initialValSet is checked in NewChain case by ValidateGenesis
+	// start a new chain
 	if state.NewChain {
-		// Create the provider client in InitGenesis for new consumer chain. CCV Handshake must be established with this client id.
+		// create the provider client in InitGenesis for new consumer chain. CCV Handshake must be established with this client id.
 		clientID, err := k.clientKeeper.CreateClient(ctx, state.ProviderClientState, state.ProviderConsensusState)
 		if err != nil {
 			panic(err)
 		}
 
-		// Set default value for valset update ID
-		k.SetHeightValsetUpdateID(ctx, uint64(ctx.BlockHeight()), uint64(0))
 		// set provider client id.
 		k.SetProviderClientID(ctx, clientID)
-	} else {
-		// verify that latest consensus state on provider client matches the initial validator set of restarted chain
-		// thus, IBC genesis MUST run before CCV consumer genesis
-		consState, ok := k.clientKeeper.GetLatestClientConsensusState(ctx, state.ProviderClientId)
-		if !ok {
-			panic("consensus state for provider client not found. MUST run IBC genesis before CCV consumer genesis")
-		}
-		tmConsState, ok := consState.(*ibctmtypes.ConsensusState)
-		if !ok {
-			panic(fmt.Sprintf("consensus state has wrong type. expected: %T, got: %T", &ibctmtypes.ConsensusState{}, consState))
-		}
 
-		// ensure that initial validator set is same as initial consensus state on provider client.
-		// this will be verified by provider module on channel handshake.
-		vals, err := tmtypes.PB2TM.ValidatorUpdates(state.InitialValSet)
-		if err != nil {
+		// set default value for valset update ID
+		k.SetHeightValsetUpdateID(ctx, uint64(ctx.BlockHeight()), uint64(0))
+	} else {
+		// verify genesis initial valset against the latest consensus state
+		// IBC genesis MUST run before CCV consumer genesis
+		if err := k.verifyGenesisInitValset(ctx, state); err != nil {
 			panic(err)
 		}
-		valSet := tmtypes.NewValidatorSet(vals)
+		// chain restarts without the CCV channel established
+		if state.ProviderChannelId == "" {
+			k.SetPendingSlashRequests(ctx, state.PendingSlashRequests)
+
+			// chain restarts with the CCV channel established
+		} else {
+			// set provider channel ID
+			k.SetProviderChannel(ctx, state.ProviderChannelId)
+			// set all unbonding sequences
+			for _, mp := range state.MaturingPackets {
+				k.SetPacketMaturityTime(ctx, mp.VscId, mp.MaturityTime)
+			}
+			// set outstanding downtime slashing requests
+			for _, od := range state.OutstandingDowntimeSlashing {
+				consAddr, err := sdk.ConsAddressFromBech32(od.ValidatorConsensusAddress)
+				if err != nil {
+					panic(err)
+				}
+				k.SetOutstandingDowntime(ctx, consAddr)
+			}
+			// set last transmission block height
+			err := k.SetLastTransmissionBlockHeight(ctx, state.LastTransmissionBlockHeight)
```

Comment:
Does this code work for a zero height restart? Does it even need to work for a zero height restart? note sure...